### PR TITLE
feat(stack reorder): Implement internal Reorder API

### PR DESCRIPTION
**Pull request by travigd**

### CODE REVIEW

File: cmd/av/stack_branch.go [hunk 1]
```diff
@@ -184,7 +184,17 @@ func stackBranchMove(
 		return errors.Errorf("cannot rename branch to itself")
 	}
 
-	currentMeta, _ := tx.Branch(oldBranch)
+	currentMeta, ok := tx.Branch(oldBranch)
```
Comment (travigd) [inline cmd/av/stack_branch.go:187]: Unrelated fix

File: internal/meta/jsonfiledb/readtx.go [hunk 2]
```diff
@@ -15,12 +15,15 @@ func (tx *readTx) Repository() (meta.Repository, bool) {
 	return tx.state.RepositoryState, tx.state.RepositoryState.ID != ""
 }
 
-func (tx *readTx) Branch(name string) (branch meta.Branch, ok bool) {
-	branch, ok = tx.state.BranchState[name]
-	if !ok {
+func (tx *readTx) Branch(name string) (meta.Branch, bool) {
```
Comment (travigd) [inline internal/meta/jsonfiledb/readtx.go:18]: Also unrelated, was trying to figure out bug above, but thought it fine to leave this in

File: internal/reorder/cmds.go [hunk 3]
```diff
@@ -28,9 +32,53 @@ func (c *Context) Print(a ...any) {
...
 	// The name of the current branch in the reorder operation.
-	Branch string
+	Branch string `json:"branch"`
+	// The sequence of commands to be executed.
+	// NOTE: we handle marshalling/unmarshalling in the MarshalJSON/UnmarshalJSON methods.
+	Commands []Cmd `json:"-"`
+}
+
+func (s *State) MarshalJSON() ([]byte, error) {
+	fmt.Println("marshalling state")
```
Comment (draftcode) [inline internal/reorder/cmds.go:44]: Debug statement.

File: internal/reorder/cmds.go [hunk 4]
```diff
@@ -28,9 +32,53 @@ func (c *Context) Print(a ...any) {
...
+
+func (s *State) UnmarshalJSON(data []byte) error {
+	// Create Alias type to avoid copying UnmarshalJSON method (and avoid infinite recursion).
+	type Alias State
+	var aux struct {
+		Commands []string `json:"commands"`
+		Alias
+	}
+	if err := json.Unmarshal(data, &aux); err != nil {
+		fmt.Printf("failed to unmarshal state: %v\n", err)
```
Comment (draftcode) [inline internal/reorder/cmds.go:68]: Debug statement

File: internal/reorder/cmds_test.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+			StackBranchCmd{Name: "one", Trunk: "main"},
+			PickCmd{"abcd"},
+			StackBranchCmd{Name: "two", Parent: "one"},
+			PickCmd{"efgh"},
+		},
+	}
+
+	serialized, err := json.Marshal(state)
+	require.NoError(t, err, "failed to serialize state")
+	fmt.Println(string(serialized))
```
Comment (draftcode) [inline internal/reorder/cmds_test.go:24]: Debug statement. If you want, use t.Log

File: internal/reorder/stackbranch.go [hunk 6]
```diff
@@ -19,11 +21,60 @@ type StackBranchCmd struct {
...
+		parentState.Name, headCommit, _ = strings.Cut(b.Trunk, "@")
+		parentState.Trunk = true
+	} else {
+		// We assume the parent branch (if not set manually) is the previous branch
+		// in the reorder operation.
+		if b.Parent == "" {
+			b.Parent = ctx.State.Branch
+		}
+		if b.Parent == "" {
+			return ErrInvalidCmd{
```
Comment (draftcode) [inline internal/reorder/stackbranch.go:49]: Hm. I wonder if we can add a special check on this before start running commands. Optional, though.

**Review by draftcode — APPROVED** [review]

File: internal/reorder/cmds.go (hunk 4 above)
Comment (travigd) [reply internal/reorder/cmds.go:68]: ty for catching!

File: internal/reorder/stackbranch.go (hunk 6 above)
Comment (travigd) [reply internal/reorder/stackbranch.go:49]: Yeah, it's possible but a little more complex than I wanted to do here. It would require a step between parsing and executing. Maybe we can add it in the future.
